src/cmd/sign.js

```diff
@@ -57,113 +57,115 @@ export default function sign(
-            return signingResult;
-          }
-        })
-        .then((signingResult) => {
-          // All information about the downloaded files would have
-          // already been logged by signAddon().
-          if (signingResult.success) {
-            log.info(`Extension ID: ${signingResult.id}`);
-            log.info('SUCCESS');
-          } else {
-            log.info('FAIL');
-          }
-          return signingResult;
-        });
+    async function (tmpDir) {
+      await prepareArtifactsDir(artifactsDir);
+
+      let manifestData;
+
+      if (preValidatedManifest) {
+        manifestData = preValidatedManifest;
+      } else {
+        manifestData = await getValidatedManifest(sourceDir);
+      }
+
+      let [buildResult, idFromSourceDir] = await Promise.all([
+        build({sourceDir, artifactsDir: tmpDir.path()}, {manifestData}),
+        getIdFromSourceDir(sourceDir),
+      ]);
+
+      const manifestId = getManifestId(manifestData);
+
+      if (id && manifestId) {
+        throw new WebExtError(
+          `Cannot set custom ID ${id} because manifest.json ` +
+          `declares ID ${manifestId}`);
+      }
+
+      if (manifestId) {
+        id = manifestId;
+      }
+
+      if (!id && idFromSourceDir) {
+        log.info(
+          'Using previously auto-generated extension ID: ' +
+          `${idFromSourceDir}`);
+        id = idFromSourceDir;
+      }
+
+      if (!id) {
+        log.warn('No extension ID specified (it will be auto-generated)');
+      }
+
+      let signingResult = await signAddon({
+        apiKey,
+        apiSecret,
+        apiUrlPrefix,
+        timeout,
+        verbose,
+        id,
+        xpiPath: buildResult.extensionPath,
+        version: manifestData.version,
+        downloadDir: artifactsDir,
+      });
+
+      if (signingResult.id) {
+        await saveIdToSourceDir(sourceDir, signingResult.id);
+      }
+
+      // All information about the downloaded files would have
+      // already been logged by signAddon().
+      if (signingResult.success) {
+        log.info(`Extension ID: ${signingResult.id}`);
+        log.info('SUCCESS');
+      } else {
+        log.info('FAIL');
+      }
+
+      return signingResult;
     }
   );
 }
 
 
-export function getIdFromSourceDir(sourceDir: string): Promise<string> {
+export async function getIdFromSourceDir(
+  sourceDir: string
+): Promise<string|void> {
   const filePath = path.join(sourceDir, extensionIdFile);
-  return fs.readFile(filePath)
-    .then((content) => {
-      let lines = content.toString().split('\n');
-      lines = lines.filter((line) => {
-        line = line.trim();
-        if (line && !line.startsWith('#')) {
-          return line;
-        }
-      });
-      let id = lines[0];
-      log.debug(`Found extension ID ${id} in ${filePath}`);
-      if (!id) {
-        throw new WebExtError(`No ID found in extension ID file ${filePath}`);
+
+  try {
+    const content = await fs.readFile(filePath);
```

Comment:
@kumar303 yep, we can do it, but we need to move `content` declaration out of the try/catch block, or it will not be available to the rest of the code in the function.

I've applied this change in 777a8bf0